### ISSUE TITLE
Invoke-IcingaCheckCPU mostly does not show values for registered -TimeIndexes

### ISSUE THREAD

**Issue by git-kick** [Icinga/icinga-powershell-framework#289]
After setting `-TimeIndexes` using `Register-IcingaServiceCheck`, `Invoke-IcingaCheckCPU` is expected to show values for different time intervals.
I am observing the correct output only in rare situations, mostly `Invoke-IcingaCheckCPU` shows only the current values, like no `-TimeIntervals` were registered.
The same was observed for `Invoke-IcingaCheckMemory`.

## Expected Behavior

After setting `-TimeIndexes` using `Register-IcingaServiceCheck`, `Invoke-IcingaCheckCPU` should always show values for all registered `-TimeIndexes`.

## Current Behavior

`Invoke-IcingaCheckCPU` mostly shows only current values. Sometimes, the output correctly differentiates the `-TimeIndexes`.
By trial & error, I observed
- Windows reboot mostly helps to get one correct sample.
- Starting a new Windows PowerShell session sometimes helps to get a correct sample.
- Restarting the icinga2 or icingapowershell service does not improve behavior.
- Mostly, the second call of `Invoke-IcingaCheckCPU` (after a successful one) shows the failure again. 

## Steps to Reproduce (for bugs)

1. `PS C:\Users\Administrator> Use-Icinga`

2. 
```
PS C:\Users\Administrator> Show-IcingaRegisteredServiceChecks
[Notice]: Service Id: 5275219864641021224811420224776891459631192206
[Notice]:
Name                           Value
----                           -----
CheckCommand                   Invoke-IcingaCheckCPU
Interval                       30
Arguments
Id                             5275219864641021224811420224776891459631192206
TimeIndexes                    {1, 5, 15}
```

3. 
```
PS C:\Users\Administrator> Invoke-IcingaCheckCPU
[OK] CPU Load
| 'core_0_5'=52.603304%;;;0;100 'core_0_1'=60.34506%;;;0;100 'core_0_15'=48.698308%;;;0;100 'core_0'=100%;;;0;100 'core_total_5'=52.603304%;;;0;100 'core_total_1'=60.34506%;;;0;100 'core_total_15'=48.698308%;;;0;100 'core_total'=100%;;;0;10
0
0
PS C:\Users\Administrator> Invoke-IcingaCheckCPU
[OK] CPU Load
| 'core_0'=41.64491%;;;0;100 'core_total'=41.64491%;;;0;100
0
```

First result is as expected; the second one is considered wrong. Both calls were done within some seconds. I can reproduce this "inconsistency" in my environment by restarting Windows and/or Windows PowerShell.
The first (successful) call runs significantly longer than the second one (and subsequent).

## Context

I am setting up a mixed Linux/Windows environment, using Icinga for Windows. Obviously, it makes sense to use `-ThresholdIntervals` instead of current values for alarming on CPU an Memory consumption (like Linux standard checks do). 
While `-ThresholdIntervals` are supported since V1.5.0 of Icinga for Windows, they currently do not work for me, supposedly due to the bug reported here.

## Your Environment

* PowerShell Version used (`$PSVersionTable.PSVersion`): 5.1.14393

* Operating System and version (`Get-IcingaWindowsInformation Win32_OperatingSystem | Select-Object Version, BuildNumber, Caption`): 10.0.14393 14393       Microsoft Windows Server 2016 Standard

* I reproduced the behavior like above on a Windows 10 Pro machine. On a Hyper-V machine running Windows 10 Enterprise Evaluation, I could _not_ achieve one single successful call of `Invoke-IcingaCheckCPU`

**Comment by LordHepipud** [comment]
Thank you for the report. Can you play try the latest 1.5.0 support branch? https://github.com/Icinga/icinga-powershell-framework/tree/support/1.5

There was a fix applied for another issue, which might resolve this one. This error should right now only occur, if you are executing the check within the same PowerShell session. Using another session it should work.

The fix in 1.5.0 branch will resolve this and be released with 1.5.1 next week.

**Comment by git-kick** [comment]
Thank you for handling the issue and sorry for delayed response (due to holidays).

I tried the 1.5 support branch as of now: [https://github.com/Icinga/icinga-powershell-framework/tree/support/1.5]
Even after reboot and with a new PowerShell session, `Invoke-IcingaCheckCPU` does not deliver the expected output, i.e. I am always getting only current values instead of {1,5,15} intervals.

Perhaps this is due to later modifications on support/1.5 within the 6 past days; could you please point me to a current branch having the fix?

**Comment by LordHepipud** [comment]
Thank you for the additional input and the long delay. I played around with this feature and I'm unable to reproduce it. The performance data is collected properly on my test machines.

Could you please provide me with overview details for the following commands:

```powershell
Get-IcingaBackgroundDaemons;
Get-ChildItem -Path "$(Get-IcingaCacheDir)\sc_daemon" -Recurse;
Get-Content -Path "$(Get-IcingaCacheDir)\sc_daemon\checkresult\Invoke-IcingaCheckCPU.json" -Raw;
Get-Content -Path "$(Get-IcingaCacheDir)\sc_daemon\checkresult_store\Invoke-IcingaCheckCPU.json" -Raw;
Get-Service icingapowershell;
Test-IcingaAgent;
```

**Comment by git-kick** [comment]
Uh, so it seems something goes wrong specifically on my Windows machines. Here comes the session log including the listed commands:
```
PS C:\Users\Administrator> Use-Icinga
PS C:\Users\Administrator> Show-IcingaRegisteredServiceChecks
[Notice]: Service Id: 5275219864641021224811420224776891459631192206
[Notice]:
Name                           Value
----                           -----
CheckCommand                   Invoke-IcingaCheckCPU
Interval                       30
Arguments
Id                             5275219864641021224811420224776891459631192206
TimeIndexes                    {1, 5, 15}



[Notice]: Service Id: 225104251186107973819916186121177195202524160180123241
[Notice]:
Name                           Value
----                           -----
CheckCommand                   Invoke-IcingaCheckMemory
Interval                       30
Arguments
Id                             225104251186107973819916186121177195202524160180123241
TimeIndexes                    {1, 5, 15}



PS C:\Users\Administrator> Invoke-IcingaCheckCPU
[OK] CPU Load
| 'core_0'=46.03269%;;;0;100 'core_total'=46.03269%;;;0;100
0
PS C:\Users\Administrator> Get-IcingaBackgroundDaemons

Name                           Value
----                           -----
Start-IcingaServiceCheckDaemon {}


PS C:\Users\Administrator> Get-ChildItem -Path "$(Get-IcingaCacheDir)\sc_daemon" -Recurse


    Verzeichnis: C:\Program Files\WindowsPowerShell\Modules\icinga-powershell-framework\cache\sc_daemon


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       08.06.2021     09:17                checkresult
d-----       08.06.2021     09:17                checkresult_store


    Verzeichnis: C:\Program Files\WindowsPowerShell\Modules\icinga-powershell-framework\cache\sc_daemon\checkresult


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       07.08.2021     07:47            479 Invoke-IcingaCheckCPU.json
-a----       07.08.2021     07:47            483 Invoke-IcingaCheckMemory.json


    Verzeichnis: C:\Program Files\WindowsPowerShell\Modules\icinga-powershell-framework\cache\sc_daemon\checkresult_store


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       07.08.2021     07:47           4851 Invoke-IcingaCheckCPU.json
-a----       07.08.2021     07:47           5241 Invoke-IcingaCheckMemory.json


PS C:\Users\Administrator> Get-Content -Path "$(Get-IcingaCacheDir)\sc_daemon\checkresult\Invoke-IcingaCheckCPU.json" -Raw
{
    "Invoke-IcingaCheckCPU":  {
                                  "Core_0_5":  47.602885,
                                  "Core_Total_5":  47.603067,
                                  "Core_0_1":  50.14672,
                                  "Core_0_15":  46.512374827586206896551724138,
                                  "Core_Total_1":  50.14672,
                                  "Core_Total_15":  46.51254
                              }
}

PS C:\Users\Administrator> Get-Content -Path "$(Get-IcingaCacheDir)\sc_daemon\checkresult_store\Invoke-IcingaCheckCPU.json" -Raw
{
    "Invoke-IcingaCheckCPU":  {
                                  "Core 0":  {
                                                 "1628314494":  "41.82405",
                                                 "1628315099":  "47.55287",
                                                 "1628314940":  "47.15975",
                                                 "1628314462":  "54.65786",
                                                 "1628314717":  "45.79492",
                                                 "1628315194":  "52.05735",
                                                 "1628315003":  "45.49175",
                                                 "1628315035":  "43.29248",
                                                 "1628315131":  "48.06833",
                                                 "1628315258":  "48.42775",
                                                 "1628314526":  "47.77373",
                                                 "1628314654":  "46.42324",
                                                 "1628314749":  "45.68875",
                                                 "1628314367":  "47.59111",
                                                 "1628314781":  "45.25503",
                                                 "1628315163":  "48.31745",
                                                 "1628314971":  "47.95582",
                                                 "1628314844":  "42.07737",
                                                 "1628314813":  "45.60719",
                                                 "1628314558":  "47.49758",
                                                 "1628315067":  "42.99936",
                                                 "1628314908":  "47.11418",
                                                 "1628314590":  "47.85268",
                                                 "1628314399":  "44.51265",
                                                 "1628314685":  "46.71107",
                                                 "1628314622":  "46.81583",
                                                 "1628314430":  "40.12539",
                                                 "1628315226":  "51.86569",
                                                 "1628314876":  "42.34764"
                                             },
                                  "Core Total":  {
                                                     "1628314494":  "41.85271",
                                                     "1628315099":  "47.55287",
                                                     "1628314940":  "47.15975",
                                                     "1628314462":  "54.65786",
                                                     "1628314717":  "45.79492",
                                                     "1628315194":  "52.03374",
                                                     "1628315003":  "45.49175",
                                                     "1628315035":  "43.29248",
                                                     "1628315131":  "48.09376",
                                                     "1628315258":  "48.42775",
                                                     "1628314526":  "47.74804",
                                                     "1628314654":  "46.42324",
                                                     "1628314749":  "45.68875",
                                                     "1628314367":  "47.59111",
                                                     "1628314781":  "45.25503",
                                                     "1628315163":  "48.31745",
                                                     "1628314971":  "47.95582",
                                                     "1628314844":  "42.07737",
                                                     "1628314813":  "45.60719",
                                                     "1628314558":  "47.49758",
                                                     "1628315067":  "42.99936",
                                                     "1628314908":  "47.11418",
                                                     "1628314590":  "47.85268",
                                                     "1628314399":  "44.51265",
                                                     "1628314685":  "46.71107",
                                                     "1628314622":  "46.81583",
                                                     "1628314430":  "40.12539",
                                                     "1628315226":  "51.86569",
                                                     "1628314876":  "42.34764"
                                                 }
                              }
}

PS C:\Users\Administrator> Get-Service icingapowershell

Status   Name               DisplayName
------   ----               -----------
Running  icingapowershell   Icinga PowerShell Service


PS C:\Users\Administrator> Test-IcingaAgent
[Passed]: Icinga Agent service is installed
[Passed]: The specified user "NT AUTHORITY\NetworkService" is allowed to run as service
[Passed]: Directory "C:\ProgramData\icinga2\etc" is accessible and writeable by the Icinga Service User "NT AUTHORITY\NetworkService"
[Passed]: Directory "C:\ProgramData\icinga2\var" is accessible and writeable by the Icinga Service User "NT AUTHORITY\NetworkService"
[Passed]: Directory "C:\Program Files\WindowsPowerShell\Modules\icinga-powershell-framework\cache" is accessible and writeable by the Icinga Service User "NT AUTHORITY\NetworkService"
[Passed]: Icinga Agent configuration is valid
[Passed]: Icinga Agent debug log is disabled
PS C:\Users\Administrator>
```
I'm happy to provide more if needed - thank you.

**Comment by LordHepipud** [comment]
Thank you for all the input. Everything seems just fine - the data is there and should be added. There is nothing obvious wrong or missing.

Do you have the Icinga Framework Code Cache enabled?

```powershell
Get-IcingaFrameworkCodeCache
```

If it returns `True`, could you please run

```powershell
Write-IcingaFrameworkCodeCache
```

Just in case the update process did not update the cache file properly and updates were missing.

If it returns `False`, just leave it as it is for the moment.

**Comment by git-kick** [comment]
````
PS C:\Users\Administrator> Get-IcingaFrameworkCodeCache
False
````
So I did not change anything further for the moment.

**Comment by LordHepipud** [comment]
This is interesting. Are both, Icinga Agent and Icinga PowerShell service running with the same user? It is recommended that the service users are identical, because right now we only support this behavior properly.

`Test-IcingaAgent` will lookup the service user the Icinga Agent is running with and test if all files and directories are accessable. As the files are present, it means the PowerShell service is actually able to write the data down. Now when the Icinga Agent executes the plugin, it will read the cached files from the location we checked above.

As `Test-IcingaAgent` is not throwing any errors, this would mean that the files can be accessed from this this.

Can you please share the following content of the output?

```powershell
Get-Module -ListAvailable -Name 'icinga-powershell-*';
```

This will print all modules installed from Icinga for Windows. Maybe we have some duplicates in here or something else weird going on.

**Comment by git-kick** [comment]
Yes, I kept the same user for both services. Initially both services were configured with user `NT Authority\NetworkService`. Then, I changed both services to the other (admin) user. The Icinga modules are looking fine for me:
```
PS C:\WINDOWS\system32> Get-Module -ListAvailable -Name 'icinga-powershell-*';

    Verzeichnis: C:\Program Files\WindowsPowerShell\Modules

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     1.5.2      icinga-powershell-framework         {Use-Icinga, Invoke-IcingaCommand, Import-IcingaLib, Get-IcingaFrameworkCodeCacheFile...}
Script     1.5.0      icinga-powershell-plugins           {Get-IcingaMemoryConfiguredVoltage, Get-IcingaDirectoryCreationTimeEqual, Get-IcingaDirectorySizeSmallerThan, Add-IcingaProcessPerfData...}
```

**Comment by git-kick** [comment]
I reproduced the Aug 7 observation on another (virtual) machine: by changing the login-account for both `icinga2` and `icingapowershell` services, I got a well-working state. 
This time, I chose the `Local System account` and after modifying ACLs as suggested by `Test-IcingaAgent`, it worked.
So, I am pretty sure that the issue is due to some account dependency. `NT Authority\NetworkService` - which is proposed in [https://icinga.com/docs/icinga-for-windows/latest/doc/installation/04-Icinga-Agent-Wizard/] - seems to be a bad choice.

While working on that, I found a minor inconsistency concerning `-TimeInterval`:
Icinga online doc, e.g. [https://icinga.com/docs/icinga-for-windows/latest/plugins/doc/plugins/04-Invoke-IcingaCheckCPU/] says `-TimeInterval` has Type `Object`.
But setting `icingacheckmemory_object_thresholdinterval` in `services.conf` has no effect.
Using `Get-IcingaCheckCommandConfig -IcingaConfig`, I found that `-TimeInterval` is declared as `String`.
So, setting `icingacheckmemory_string_thresholdinterval` does the trick.

Please let me know whether you think this is worth a new issue, and where I should report it (if not here).

**Comment by git-kick** [comment]
Since it did _not_ work with service user `LocalSystem` on my physical Win 10 machine, I executed a test series with four users for both services and found:
`.\Uli` (a local admin user): works ok.
`LocalSystem`: service icingapowershell does not start (error 1067).
`NT AUTHORITY\LocalService`: service icingapowershell does not start (fails silently).
`NT AUTHORITY\NetworkService`: service icingapowershell does not start (fails silently).
In all cases, `Set-IcingaServiceUser` and `Restart-IcingaService` are running smoothly without error messages.

[Set-IcingaService-User.log](https://github.com/Icinga/icinga-powershell-framework/files/6990366/Set-IcingaService-User.log)

**Comment by LordHepipud** [comment]
We might have another issue for this: #336

I have however no idea on why this error ocurres. If you are still running 1.6.0:

* Update to the current master
* Run `Test-IcingaForWindowsService`


With Icinga for Windows 1.6.0 you will require to load the .psd1 file instead of the .psm1 file from the PowerShell module. The command hints you on some problems and allows you to resolve them with

```powershell
Test-IcingaForWindowsService -ResolveProblems
```

If you are running Icinga 2.13.0, then there will no longer be a debug log or icinga2 log available. Everything is written into the EventLog by default now. I'm trying to get something worked out to simply read the logs into that as before.

**Comment by git-kick** [comment]
Thank you for pointing me to [#336](https://github.com/Icinga/icinga-powershell-framework/issues/336). It is somehow related, but I never saw these 'UNKNOWN' results in my case. Instead, I always got only the (correct) current value instead of expected four values (current,1m,5m,15m). Also, I no longer have any `-TimeIndexes` issues after working around the service user issue as described above.
So, does it make sense to alter the title of this #289 to something like **"Service icingapowershell not running properly with recommended user accounts"**?  

I updated to the current master. Icinga2 checks (ping, cluster, icinga) are up and running, icingapowershell checks are not (as assumed). So I did
```
PS C:\Tools\icinga2> Test-IcingaForWindowsService
[Passed]: Your service installation is not affected by IWKB000009
[Failed]: Your service installation is referring "icinga-powershell-framework.psm1". This is deprecated and has to be changed to "icinga-powershell-framework.psd1". Run this Cmdlet with "-ResolveProblems" to fix it.
False
PS C:\Tools\icinga2> Test-IcingaForWindowsService -ResolveProblems
[Passed]: Your service installation is not affected by IWKB000009
[Warning]: Your service installation is referring to "icinga-powershell-framework.psm1" for module imports. Trying to resolve the problem.
[Passed]: Your service installation is now properly referring to "icinga-powershell-framework.psd1" for module imports.
True
PS C:\Tools\icinga2> Test-IcingaForWindowsService
[Passed]: Your service installation is not affected by IWKB000009
[Failed]: Your service installation is referring "icinga-powershell-framework.psm1". This is deprecated and has to be changed to "icinga-powershell-framework.psd1". Run this Cmdlet with "-ResolveProblems" to fix it.
False
```
which is strange because `-ResolveProblems` ran ok but obviously did not really fix the psm1/psd1 problem.
I had a look into `Test-IcingaForWindowsService.psm1`, and it ends up in setting `$PreparedServicePath` to "...psm1" which is probably not intended, because this does not change anything. So I added some `-replace` to Test-`IcingaForWindowsService.psm1` just for testing:
```
    [string]$PreparedServicePath = [string]::Format(
        '\"{0}\" \"{1}\"',
        $ServiceData.FullPath,
        (Get-IcingaPowerShellModuleFile)
    ) -replace 'psm1','psd1';
```
After running `Test-IcingaForWindowsService -ResolveProblems` , `Test-IcingaForWindowsService` does not complain anymore.
**But still, service 'icingapowershell' does not start properly.**
 
Thank you for the icinga-log/EventLog remark. Did I get it right that you are aiming to make `Read-IcingaAgent[Debug]LogFile` work again (reading from EventLog)? Otherwise, I would suggest to mark those two cmdlets 'deprecated'.

**Comment by LordHepipud** [comment]
Sorry for the late response - yes, the plan is to include EventLog Monitoring in case the feature is used directly from the EventLog. Not sure if I can make it with 1.7.0 or 1.8.0.

The other issues should already be resolved with v1.7.0 and v1.6.1. Did I miss anything else in this matter?

I'm still wondering on why sometimes the regular user does not seem to work.

**Comment by git-kick** [comment]
Thank you for picking up this issue again, and for the answer concerning EventLog.

The original issue (-TimeIndexes not working properly) seems to be an implication of the service user issue. As we did not observe the service user issue in production (only on some test machines), **I suggest to close this issue**.

Perhaps you want to look at **one remaining (minor) point** - I'll be happy to create a new issue for that, if appropriate:
> While working on that, I found a minor inconsistency concerning -TimeInterval:
> Icinga online doc, e.g. [https://icinga.com/docs/icinga-for-windows/latest/plugins/doc/plugins/04-Invoke-IcingaCheckCPU/] says -TimeInterval has Type Object.
> But setting icingacheckmemory_object_thresholdinterval in services.conf has no effect.
> Using Get-IcingaCheckCommandConfig -IcingaConfig, I found that -TimeInterval is declared as String.
> So, setting icingacheckmemory_string_thresholdinterval does the trick.

**Comment by LordHepipud** [comment]
Thank you for the feedback - a separate issue for this would be great. A fix might ship with v1.7.0 already then.

Thank you a lot!
